rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,477 @@
+- Validate the `no_tls_reason` field of JSON output to ensure that the edges are mTLS'ed.
+
+**linkerd routes**
+- Check if application pods are up and running
+- Issue `linkerd routes` cmd
+- Validate the returned output by counting instances of desired route substrings using `strings.Count`
+  
+**3. Ingress**
+
+It is essential to test ingress to ensure that the ingress controller re-writes incoming headers to the internal service name. This process ensures that service discovery is working. This test shall include deploying various types of ingress controllers (Nginx, Traefik and Ambassador) :
+
+- Deploy and verify if the ingress controller is up and running. The YAML for various ingress controllers shall be provided along with the test tool. 
+- Deploy the ingress resource. An example from the docs:
+```yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: web-ingress
+  namespace: emojivoto
+  annotations:
+    kubernetes.io/ingress.class: "nginx"
+    nginx.ingress.kubernetes.io/configuration-snippet: |
+      proxy_set_header l5d-dst-override $service_name.$namespace.svc.cluster.local:$service_port;
+      grpc_set_header l5d-dst-override $service_name.$namespace.svc.cluster.local:$service_port;
+
+spec:
+  rules:
+  - host: example.com
+    http:
+      paths:
+      - backend:
+          serviceName: web-svc
+          servicePort: 8080
+```
+- Obtain external IP of the controller. For e.g :
+```bash
+$ kubectl get svc --all-namespaces \
+  -l app=nginx-ingress,component=controller \
+  -o=custom-columns=EXTERNAL-IP:.status.loadBalancer.ingress[0].ip
+```
+> Note: The ingress resource file and the command to obtain the external IP may change according to the ingress controller being tested.
+
+- Issue a curl cmd on the external IP to check if desired response is returned. This shall be done by wrapping the curl bash cmd as a method in Golang using `command.exec()`.
+
+**4. Data plane proxy health checks**
+
+- Issue a `check` command - `linkerd -n <ns> check --proxy -o json`
+- From the output JSON, under `"categoryName" : "linkerd-data-plane"`, verify if `result` of each check under the `checks` array shows `success`.
+- Make a `GET` request to the `linkerd-proxy` containers (of each of the pods) at the `/metrics` (Liveness probe) and `/ready` (Readiness probes) to ensure that they are reachable.
+- From the `linkerd-proxy` container of each of the pods, check for 503 errors.
+
+**5. `tap` extension API server**
+
+Some custom Kubernetes clusters don't always have the aggregation layer configured, causing the tap service which is an extension API server to fail.
+- Issue a `check` command - `linkerd -n <ns> check --pre -o json`
+- Validate the returned JSON by ensuring that the array the check with description `"can read extension-apiserver-authentication configmap"` under category `"pre-kubernetes-setup"` says `"success"`
+
+**6. Retries and Timeouts**
```

Comment:
Something to think about is that instead of relying on a property like `effective_success` to determine the test assertion, which has proven to be flakey, it will be nice if there is a way to record/observe the occurrence of retries on the server side, and timeouts on the client side.